vllm/v1/guided_decoding/__init__.py

```diff
@@ -0,0 +1,195 @@
+# SPDX-License-Identifier: Apache-2.0
+from __future__ import annotations
+
+import copy
+import enum
+import threading
+from concurrent.futures import ThreadPoolExecutor
+from dataclasses import dataclass, field
+from typing import TYPE_CHECKING, Dict, List, Optional, Set, Tuple
+
+import torch
+import xgrammar as xgr
+
+from vllm.config import VllmConfig
+from vllm.logger import init_logger
+from vllm.transformers_utils.tokenizer_group import init_tokenizer_from_configs
+
+if TYPE_CHECKING:
+    from vllm.v1.request import Request
+
+logger = init_logger(__name__)
+
+
+class GuidedDecodingOptions(enum.Enum):
+    json = enum.auto()
+    json_object = enum.auto()
+    regex = enum.auto()
+    grammar = enum.auto()
+    choice = enum.auto()
+
+
+GuidedDecodingKey = Tuple[GuidedDecodingOptions, str]
+MAX_ROLLBACK_TOKENS = 100
```

Comment:
Where does `100` come from?

Reply:
I was just using a default to test out, but I will probably revisit this afterwards

Reply:
It should be safe to use max-num-speculative-tokens + 1, if that is known at startup time

Reply:
After second thought, shouldn't this be related to request preemption and re-computation?

Reply:
I've dropped this constant and its use because we don't actually support any rollback yet, so leaving the partial implementation here is confusing.